### PR TITLE
Allow mapping of ZMK Bluetooth actions, ignore pycache

### PR DESCRIPTION
**Pull request by minusfive**
This PR allows mapping of Bluetooth actions to enable use of glyphs (e.g. Bluetooth logo/icons) on `&bt` behaviours.

E.g. drawn locally using my config https://github.com/minusfive/zmk-config-zen-2

<details>
  <summary>YAML</summary>

```yaml
layout: {qmk_keyboard: corne_rotated, qmk_layout: LAYOUT_split_3x6_3}
layers:
  L1:
  - $$mdi:keyboard-esc$$
  - Q
  - W
  - F
  - P
  - B
  - J
  - L
  - U
  - Y
  - {t: '''', s: '"'}
  - {t: '`', s: '~'}
  - $$mdi:backspace$$
  - A
  - R
  - S
  - T
  - G
  - M
  - N
  - E
  - I
  - O
  - $$mdi:keyboard-return$$
  - {t: '$$mdi:keyboard-tab$$', h: Meh}
  - {t: Z, h: '$$mdi:apple-keyboard-control$$'}
  - {t: X, h: '$$mdi:apple-keyboard-option$$'}
  - C
  - D
  - V
  - K
  - H
  - {t: ',', s: <}
  - {t: ., h: '$$mdi:apple-keyboard-option$$', s: '>'}
  - {t: /, h: '$$mdi:apple-keyboard-control$$', s: '?'}
  - {t: \, h: Meh, s: '|'}
  - {t: '[', h: '$$mdi:apple-keyboard-command$$', s: '{'}
  - {t: '$$mdi:keyboard-space$$', h: '$$mdi:apple-keyboard-shift$$'}
  - {t: ']', h: L2, s: '}'}
  - {t: '-', h: L2, s: _}
  - {t: '=', h: '$$mdi:apple-keyboard-shift$$', s: +}
  - {t: ;, h: '$$mdi:apple-keyboard-command$$', s: ':'}
  L2:
  - L3
  - {t: '1', s: '!'}
  - {t: '2', s: '@'}
  - {t: '3', s: '#'}
  - {t: '4', s: $}
  - {t: '$$mdi:border-none-variant$$', type: trans}
  - $$mdi:volume-off$$
  - $$mdi:volume-low$$
  - $$mdi:volume-high$$
  - $$mdi:brightness-7$$
  - $$mdi:brightness-5$$
  - L3
  - {t: '$$mdi:border-none-variant$$', type: trans}
  - {t: '5', s: '%'}
  - {t: '6', s: ^}
  - {t: '7', s: '&'}
  - {t: '8', s: '*'}
  - $$mdi:apple-keyboard-caps$$
  - $$mdi:apple-keyboard-caps$$
  - $$mdi:arrow-left-bold$$
  - $$mdi:arrow-up-bold$$
  - $$mdi:arrow-down-bold$$
  - $$mdi:arrow-right-bold$$
  - {t: '$$mdi:border-none-variant$$', type: trans}
  - {t: '$$mdi:border-none-variant$$', type: trans}
  - {t: '$$mdi:border-none-variant$$', type: trans}
  - {t: '$$mdi:border-none-variant$$', type: trans}
  - {t: '9', s: (}
  - {t: '0', s: )}
  - {t: '$$mdi:border-none-variant$$', type: trans}
  - $$mdi:play-pause$$
  - $$mdi:skip-previous$$
  - $$mdi:skip-next$$
  - {t: '$$mdi:border-none-variant$$', type: trans}
  - {t: '$$mdi:border-none-variant$$', type: trans}
  - {t: '$$mdi:border-none-variant$$', type: trans}
  - {t: '$$mdi:border-none-variant$$', type: trans}
  - {t: '$$mdi:border-none-variant$$', type: trans}
  - {t: ']', s: '}', type: held}
  - {t: '-', s: _}
  - {t: '$$mdi:border-none-variant$$', type: trans}
  - {t: '$$mdi:border-none-variant$$', type: trans}
  L3:
  - $$mdi:bluetooth-off$$
  - {t: '$$mdi:bluetooth-connect$$', h: '0'}
  - {t: '$$mdi:bluetooth-connect$$', h: '1'}
  - {t: '$$mdi:bluetooth-connect$$', h: '2'}
  - {t: '$$mdi:bluetooth-connect$$', h: '3'}
  - {t: '$$mdi:bluetooth-connect$$', h: '4'}
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - ''
  - L1
  - L1
  - ''
  - ''
```

</details>

km

### CODE REVIEW

**Comment by caksoylar** [comment]
Thanks for the PR! I like the idea and the end result, but it bothers me a bit that this lives in `zmk_keycode_map` since that is only for proper keycodes (i.e. ones that can be used with `&kp`). I envisioned `raw_bindings_map` to hold other such mappings but it is not flexible enough, e.g. to do partial replacements of the behavior+params like you do here. Let me think about this a bit and see if there is an alternative solution.

**Comment by minusfive** [comment]
@caksoylar I hear you, was thinking along similar lines. Did it here since it’s where all behaviors currently seem to be managed, and where `&bt` itself was, and wanted it to be a small change both for ease of iteration/quick/immediate improvement, + honestly lack of time.

But was also thinking perhaps the opposite approach: rather than creating separate maps, a single `zmk_behavior_actions_key_map` (terrible name) could be useful, with a structure similar to:


```yaml
zmk_behavior_actions_key_map:
  kp:
    DOT:
      t: ‘.’
      s: ‘>’
      # …
  bt:
    BR_SEL:
      t: ‘$$mdi:bluetooth-connect$$
      h: ‘$$params[0]$$’
      # …
```

And so on. The `$$params[0]$$` bit is an idea to allow full customization of where each fragment is displayed, with standardized named tokens straight from ZMK, so the naming is consistent with what people are used to. E.g. `$$behavior$$`, `$$keycode$$`, `$$action$$`, `$$params[0]$$`, etc.

That’s just the seed of the idea, sorry, typing from my phone.

**Comment by caksoylar** [comment]
I like that idea! I would call it `zmk_behavior_param_map` and make `zmk_keycode_map` alias to `zmk_behavior_param_map.kp` etc.

If you like, we could merge this PR as is and I can refactor later to split them up, if you want to keep this quick. (I might also be "particular" about how to implement things as this is my _fun_ project 😄.) But I am also happy to let you give it a shot with rounds of feedback.

Re: the parametrization idea, I am not sure about such an approach just yet. I know there is already precedent for special string syntax for glyphs (slightly different that it is resolved in draw time, not parsing) but extensions like this have a habit of turning into bespoke domain-specific languages 😄

**Comment by minusfive** [comment]
@caksoylar if you’re OK with it, yeah, it’d be great to merge it and get the feature out!

And let’s see who gets to the refactor first ;).

I agree re: tokens, likely to get messy/hairy.

I also have some ideas to enable theming, and an alternative way of displaying transparent keys. May give those a shot at some point, will keep you in the loop.

At least I hope, hard to juggle side projects nowadays with a toddler 🤣.

File: keymap_drawer/parse.py [hunk 1]
```diff
@@ -246,8 +246,13 @@ def mapped(key: str) -> LayoutKey:
             case [ref, par] if ref in self.sticky_keys:
                 l_key = self._str_to_key(f"{self.sticky_keys[ref][0]} {par}", current_layer, key_positions)
                 return LayoutKey(tap=l_key.tap, hold=self.cfg.sticky_label, shifted=l_key.shifted)
-            case [("&out" | "&bt" | "&ext_power" | "&rgb_ug"), *pars]:
-                return LayoutKey(tap=" ".join(pars).replace("_", " ").replace(" SEL ", " "))
```
Comment (caksoylar) [inline keymap_drawer/parse.py:250]: Can you move `.replace(" SEL ", " ")` to the default processing in line 228? This way we would preserve the current behavior. (You'll probably need to run `black` after.)

**Review by caksoylar — CHANGES_REQUESTED** [review]

File: keymap_drawer/parse.py (hunk 1 above)
Comment (minusfive) [reply keymap_drawer/parse.py:250]: Will do!

Comment (minusfive) [reply keymap_drawer/parse.py:250]: @caksoylar done!

**Review by caksoylar — APPROVED** [review]

**Comment by caksoylar** [comment]
@minusfive thanks for the contribution! I didn't think through some issues with my last suggestion, so I pushed a fix to it here: 0c32be610ad36168ee9ae5fcd1d327bbac2b4e3d

Also [added a new input](https://github.com/caksoylar/keymap-drawer/commit/86d20a3e1e0e6a09706a2e802fc6d2b557144be3) to the reusable workflow so you can use the `main` branch in your drawing workflow instead of the latest release, which will allow you to utilize this change before then.

**Comment by minusfive** [comment]
@caksoylar nice, thank you! I’ve reconfigured my workflow to use the branch input and works nicely! Me README svg was redrawn correctly.

**Pull request merged**